In [1]:
import pandas as pd
melb_df = pd.read_csv('data/melb_data_fe.csv')
melb_df.head()

Suburb  Rooms   Type      Price Method SellerG        Date  Distance  \
0  Abbotsford      2  house  1480000.0      S  Biggin  2016-03-12       2.5   
1  Abbotsford      2  house  1035000.0      S  Biggin  2016-04-02       2.5   
2  Abbotsford      3  house  1465000.0     SP  Biggin  2017-04-03       2.5   
3  Abbotsford      3  house   850000.0     PI  Biggin  2017-04-03       2.5   
4  Abbotsford      4  house  1600000.0     VB  Nelson  2016-04-06       2.5   

   Postcode  Bedroom  ...  Longtitude             Regionname  Propertycount  \
0      3067        2  ...    144.9984  Northern Metropolitan           4019   
1      3067        2  ...    144.9934  Northern Metropolitan           4019   
2      3067        3  ...    144.9944  Northern Metropolitan           4019   
3      3067        3  ...    144.9969  Northern Metropolitan           4019   
4      3067        3  ...    144.9941  Northern Metropolitan           4019   

   MeanRoomsSquare AreaRatio  MonthSale  AgeBuilding WeekdaySale  StreetType  \
0            25.20 -0.231707          3           46           5          St   
1            15.80 -0.327660          4          116           5          St   
2            18.75  0.056338          4          117           0          St   
3            15.75  0.145455          4           47           0       other   
4            17.75  0.083969          4            2           2          St   

   Weekend  
0        1  
1        1  
2        0  
3        0  
4        0  

[5 rows x 26 columns]

In [2]:
display(melb_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 26 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   Suburb           13580 non-null  object 
 1   Rooms            13580 non-null  int64  
 2   Type             13580 non-null  object 
 3   Price            13580 non-null  float64
 4   Method           13580 non-null  object 
 5   SellerG          13580 non-null  object 
 6   Date             13580 non-null  object 
 7   Distance         13580 non-null  float64
 8   Postcode         13580 non-null  int64  
 9   Bedroom          13580 non-null  int64  
 10  Bathroom         13580 non-null  int64  
 11  Car              13580 non-null  int64  
 12  Landsize         13580 non-null  float64
 13  BuildingArea     13580 non-null  float64
 14  CouncilArea      12211 non-null  object 
 15  Lattitude        13580 non-null  float64
 16  Longtitude       13580 non-null  float64
 17  Regionname  

None

----

###  Задание 1.1 ###

Преобразовать столбец `Date` в формат `datetime` и выделить квартал (`quarter`) продажи объектов недвижимости. Найдите второй по популярности квартал продажи. В качестве ответа записать число объектов, проданных в этом квартале.

***Ответ - 4359***

In [3]:
melb_df['Date'] = pd.to_datetime(melb_df['Date'])
quarters = melb_df['Date'].dt.quarter
quarters.value_counts()

3    4873
2    4359
4    2329
1    2019
Name: Date, dtype: int64

----

### Задание 1.2 ###

Преобразовать все столбцы, в которых меньше 150 уникальных значений, в тип данных `category`, исключив из преобразования столбцы **`Date`**, **`Rooms`**, **`Bedroom`**, **`Bathroom`**, **`Car`**.

В качестве ответа записать результирующее количество столбцов, которые имеют тип данных `category`.

***Ответ - 10***

In [4]:
unique_list = []
for col in melb_df.columns:
    item = (col, melb_df[col].nunique(), melb_df[col].dtypes)
    unique_list.append(item)
unique_counts = pd.DataFrame(
    unique_list,
    columns=['Column_Name', 'Num_Unique', 'Type']
).sort_values(by='Num_Unique', ignore_index=True)
display(unique_counts)

Column_Name  Num_Unique            Type
0           Weekend           2           int64
1              Type           3          object
2            Method           5          object
3       WeekdaySale           7           int64
4        Regionname           8          object
5             Rooms           9           int64
6          Bathroom           9           int64
7               Car          11           int64
8        StreetType          11          object
9         MonthSale          12           int64
10          Bedroom          12           int64
11      CouncilArea          33          object
12          SellerG          50          object
13             Date          58  datetime64[ns]
14           Suburb         120          object
15      AgeBuilding         152           int64
16         Postcode         198           int64
17         Distance         202         float64
18    Propertycount         311           int64
19     BuildingArea         602         float64
20  MeanRoomsSquare        1275         float64
21         Landsize        1448         float64
22            Price        2204         float64
23        Lattitude        6503         float64
24        AreaRatio        6521         float64
25       Longtitude        7063         float64

In [5]:
cols_to_exclude = ['Date', 'Rooms', 'Bedroom', 'Bathroom', 'Car']
max_unique_count = 150
for col in melb_df.columns:
    if melb_df[col].nunique() < max_unique_count and col not in cols_to_exclude:
        melb_df[col] = melb_df[col].astype('category')
display(melb_df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13580 entries, 0 to 13579
Data columns (total 26 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   Suburb           13580 non-null  category      
 1   Rooms            13580 non-null  int64         
 2   Type             13580 non-null  category      
 3   Price            13580 non-null  float64       
 4   Method           13580 non-null  category      
 5   SellerG          13580 non-null  category      
 6   Date             13580 non-null  datetime64[ns]
 7   Distance         13580 non-null  float64       
 8   Postcode         13580 non-null  int64         
 9   Bedroom          13580 non-null  int64         
 10  Bathroom         13580 non-null  int64         
 11  Car              13580 non-null  int64         
 12  Landsize         13580 non-null  float64       
 13  BuildingArea     13580 non-null  float64       
 14  CouncilArea      12211 non-null  categ

None

# Сортировка данных в DataFrame #

## Метод sort_values() ##

### Сортировка по значениям одного столбца ###

In [6]:
melb_df.sort_values(by='Price').head(10)

Suburb  Rooms   Type     Price Method        SellerG  \
2652        Footscray      1   unit   85000.0     PI        Burnham   
1805            other      4  house  131000.0     PI          other   
7303           Albion      1   unit  145000.0     PI         Biggin   
1927           Coburg      4  house  145000.0     PI         Jellis   
7940         Hawthorn      1   unit  160000.0     VB            HAR   
12666       Brunswick      1   unit  170000.0     VB         Nelson   
8811        Footscray      1   unit  170000.0     PI        Burnham   
8504   West Footscray      1   unit  185000.0     PI            Jas   
7293           Albion      1   unit  185000.0      S  hockingstuart   
7305           Albion      2   unit  190000.0     SP        Burnham   

            Date  Distance  Postcode  Bedroom  ...  Longtitude  \
2652  2016-03-09       6.4      3011        1  ...   144.89000   
1805  2017-02-25       8.9      3162        4  ...   145.02420   
7303  2016-05-28      13.9      3020        2  ...   144.82660   
1927  2016-04-06       7.8      3058        3  ...   144.96580   
7940  2017-08-04       4.6      3122        1  ...   145.03730   
12666 2017-09-16       5.2      3056        1  ...   144.95188   
8811  2017-01-07       5.1      3011        1  ...   144.89587   
8504  2017-04-29       8.2      3012        1  ...   144.86720   
7293  2016-08-22      13.9      3020        1  ...   144.82720   
7305  2016-07-30      13.9      3020        2  ...   144.82390   

                  Regionname  Propertycount  MeanRoomsSquare AreaRatio  \
2652    Western Metropolitan           7570        42.000000  1.000000   
1805   Southern Metropolitan           2379        17.222222 -0.525994   
7303    Western Metropolitan           2185        31.500000  0.555556   
1927   Northern Metropolitan          11204        20.500000 -0.531429   
7940   Southern Metropolitan          11308        42.000000 -0.437500   
12666  Northern Metropolitan          11918        42.000000 -0.816860   
8811    Western Metropolitan           7570         8.666667 -0.071429   
8504    Western Metropolitan           5058        42.000000  1.000000   
7293    Western Metropolitan           2185        14.333333 -0.964017   
7305    Western Metropolitan           2185        25.200000  1.000000   

       MonthSale  AgeBuilding WeekdaySale  StreetType  Weekend  
2652           3            9           2          St        0  
1805           2           97           5          St        1  
7303           5           46           5          St        1  
1927           4          106           2          Rd        0  
7940           8            8           4          St        0  
12666          9           47           5          St        1  
8811           1            4           5          St        1  
8504           4           47           5          St        1  
7293           8           41           0          Rd        0  
7305           7           46           5          St        1  

[10 rows x 26 columns]

In [7]:
melb_df.sort_values(by='Date', ascending=False)

Suburb  Rooms   Type      Price Method    SellerG       Date  \
11144     Northcote      4  house  1955000.0     SP    McGrath 2017-12-08   
11217  Surrey Hills      3  house  1775000.0     PI     Jellis 2017-12-08   
11206      St Kilda      4  house  1600000.0     VB       Gary 2017-12-08   
11207    Strathmore      7  house  2000000.0     SP     Nelson 2017-12-08   
11208    Strathmore      4  house  1610000.0      S      other 2017-12-08   
...             ...    ...    ...        ...    ...        ...        ...   
1555     Camberwell      4  house  2650000.0      S     Jellis 2016-03-09   
4510       Oak Park      3  house   892000.0      S     Nelson 2016-03-09   
4388      Northcote      3  house  1200000.0      S     Nelson 2016-03-09   
6184   Surrey Hills      3  house  1205000.0      S  Fletchers 2016-01-28   
6183   Surrey Hills      2   unit   813000.0      S  Fletchers 2016-01-28   

       Distance  Postcode  Bedroom  ...  Longtitude             Regionname  \
11144       5.3      3070        4  ...   144.99375  Northern Metropolitan   
11217      10.2      3127        3  ...   145.08968  Southern Metropolitan   
11206       5.0      3182        4  ...   144.98324  Southern Metropolitan   
11207       8.2      3041        7  ...   144.90587   Western Metropolitan   
11208       8.2      3041        4  ...   144.92243   Western Metropolitan   
...         ...       ...      ...  ...         ...                    ...   
1555        7.8      3124        4  ...   145.06860  Southern Metropolitan   
4510       11.5      3046        3  ...   144.91740  Northern Metropolitan   
4388        5.5      3070        3  ...   144.99090  Northern Metropolitan   
6184       11.2      3127        3  ...   145.10060  Southern Metropolitan   
6183       11.2      3127        2  ...   145.10230  Southern Metropolitan   

       Propertycount  MeanRoomsSquare AreaRatio  MonthSale  AgeBuilding  \
11144          11364        12.600000  0.923664         12           47   
11217           5457        15.750000 -0.721854         12           47   
11206          13240        16.000000 -0.586028         12          107   
11207           3284        20.882353 -0.262721         12           18   
11208           3284        24.909091 -0.517181         12           62   
...              ...              ...       ...        ...          ...   
1555            8920        18.900000 -0.550535          3          126   
4510            2651        18.000000 -0.714609          3           46   
4388           11364        17.857143 -0.264706          3           96   
6184            5457        18.000000 -0.590909          1           46   
6183            5457        25.200000  0.076923          1           46   

      WeekdaySale  StreetType  Weekend  
11144           4          St        0  
11217           4          Rd        0  
11206           4          St        0  
11207           4          Rd        0  
11208           4          St        0  
...           ...         ...      ...  
1555            2          Rd        0  
4510            2          St        0  
4388            2          St        0  
6184            3          Rd        0  
6183            3          Rd        0  

[13580 rows x 26 columns]

### Сортировка по значениям нескольких столбцов ###

In [8]:
melb_df.sort_values(by=['Distance', 'Price']).loc[::10, ['Distance', 'Price']]

Distance      Price
11428       0.0   387000.0
10512       0.7   600000.0
5727        1.2   485000.0
8671        1.2   595000.0
5736        1.2   740000.0
...         ...        ...
12011      38.0   680000.0
10673      38.0   810000.0
13429      38.0  1155000.0
11102      41.0   650000.0
12728      45.9   650000.0

[1358 rows x 2 columns]

In [9]:
melb_df.sort_values(by=['Price', 'Distance']).loc[::10, ['Distance', 'Price']]

Distance      Price
2652        6.4    85000.0
7116        7.8   200000.0
5206       11.2   220000.0
7240        8.2   241000.0
2898       13.0   250000.0
...         ...        ...
13341       6.3  4060000.0
8591        1.5  4275000.0
1094       11.2  4700000.0
5801        6.1  5046000.0
251         6.3  5525000.0

[1358 rows x 2 columns]

### Комбинирование сортировки с фильтрацией ###

In [10]:
mask1 = melb_df['AreaRatio'] < -0.8
mask2 = melb_df['Type'] == 'townhouse'
mask3 = melb_df['SellerG'] == 'McGrath'
melb_df[mask1 & mask2 & mask3].sort_values(
    by=['Date', 'AreaRatio'],
    ascending=[True, False],
    ignore_index=True
).loc[:, ['Date', 'AreaRatio']]

Date  AreaRatio
0 2016-07-26  -0.974922
1 2016-09-24  -0.971831
2 2016-11-27  -0.953608
3 2016-12-11  -0.945946
4 2017-08-04  -0.947368
5 2017-08-04  -0.970874

----

###  Задание 2.2 ###

Произвести сортировку столбца `AreaRatio` по убыванию. При этом индексы полученной таблицы заменить на новые. Какое значение площади здания находится в строке 1558? Ответ округлить до целого числа.

***Ответ - 126***

In [11]:
melb_df.sort_values(by='AreaRatio', ascending=False, ignore_index=True)['BuildingArea'].iloc[1555:1560]

1555     53.3
1556     93.0
1557     81.0
1558    126.0
1559     72.0
Name: BuildingArea, dtype: float64

----

###  Задание 2.3 ###

Найти таунхаусы (`Type`) с количеством жилых комнат (`Rooms`) больше 2. Отсортировать полученную таблицу сначала по возрастанию числа комнат, а затем по убыванию средней площади комнат (`MeanRoomsSquare`). Индексы таблицы заменить на новые. Какая цена будет у объекта в строке 18? Ответ запиcать в виде целого числа.

***Ответ - 1300000***

In [12]:
melb_df[(melb_df['Rooms'] > 2) & (melb_df['Type'] == 'townhouse')].sort_values(
    by=['Rooms', 'MeanRoomsSquare'],
    ascending=[True, False],
    ignore_index=True
).loc[18, 'Price']

1300000.0

# Группировка данных в DataFrame #

## Метод groupby() ##

### Группировка данных по одному критерию с одной агрегацией ###

In [13]:
melb_df.groupby(by='Type').mean()

Rooms         Price   Distance     Postcode   Bedroom  Bathroom  \
Type                                                                            
house      3.260874  1.242665e+06  10.979479  3104.080643  3.229336  1.613822   
townhouse  2.837522  9.337351e+05   9.851346  3100.777379  2.814183  1.809695   
unit       1.963871  6.051275e+05   7.607391  3110.797481  1.966523  1.183295   

                Car    Landsize  BuildingArea  Lattitude  Longtitude  \
Type                                                                   
house      1.772674  617.181924    152.162553 -37.803795  144.994700   
townhouse  1.555655  279.606822    134.649710 -37.815782  144.996489   
unit       1.128936  477.314219    102.235863 -37.823710  144.996363   

           Propertycount  MeanRoomsSquare  AreaRatio  AgeBuilding  
Type                                                               
house        7259.025505        18.996731  -0.490031    55.669700  
townhouse    7094.459605        18.569847  -0.094916    26.690305  
unit         8199.280080        21.068242   0.319883    39.703016

In [14]:
melb_df.groupby('Type')['Price'].mean()

Type
house        1.242665e+06
townhouse    9.337351e+05
unit         6.051275e+05
Name: Price, dtype: float64

In [15]:
melb_df.groupby('Regionname')['Distance'].min().sort_values(ascending=False)

Regionname
Western Victoria              29.8
Eastern Victoria              25.2
Northern Victoria             21.8
South-Eastern Metropolitan    14.7
Eastern Metropolitan           7.8
Western Metropolitan           4.3
Southern Metropolitan          0.7
Northern Metropolitan          0.0
Name: Distance, dtype: float64

### Группировка данных по одному критерию с несколькими агрегациями ###

In [16]:
melb_df.groupby('MonthSale')['Price'].agg(
    ['count', 'mean', 'max']
).sort_values(by='count', ascending=False)

count          mean        max
MonthSale                                
8           1850  1.056371e+06  6500000.0
7           1835  9.314698e+05  9000000.0
5           1644  1.097807e+06  8000000.0
6           1469  1.068981e+06  7650000.0
3           1408  1.146762e+06  5600000.0
4           1246  1.050479e+06  5500000.0
9           1188  1.126349e+06  6400000.0
10           854  1.135970e+06  6250000.0
11           750  1.142503e+06  5050000.0
12           725  1.144737e+06  5700000.0
2            333  1.169051e+06  4735000.0
1            278  9.397921e+05  5200000.0

In [17]:
melb_df.groupby('MonthSale')['Price'].agg('describe')

count          mean            std       min       25%        50%  \
MonthSale                                                                       
1           278.0  9.397921e+05  577668.924214  170000.0  570500.0   795000.0   
2           333.0  1.169051e+06  671564.357417  131000.0  710000.0  1020000.0   
3          1408.0  1.146762e+06  709573.596867   85000.0  680000.0   945000.0   
4          1246.0  1.050479e+06  591892.902979  145000.0  655000.0   905500.0   
5          1644.0  1.097807e+06  668492.867996  145000.0  650000.0   905000.0   
6          1469.0  1.068981e+06  606010.069052  222000.0  660000.0   900000.0   
7          1835.0  9.314698e+05  537390.803161  190000.0  586750.0   800000.0   
8          1850.0  1.056371e+06  619617.476541  160000.0  635000.0   892000.0   
9          1188.0  1.126349e+06  608734.690742  170000.0  725000.0   980000.0   
10          854.0  1.135970e+06  692950.251627  250000.0  652625.0   950000.0   
11          750.0  1.142503e+06  703666.227892  240000.0  671750.0   950000.0   
12          725.0  1.144737e+06  706549.007161  250000.0  667500.0   938000.0   

                 75%        max  
MonthSale                        
1          1111250.0  5200000.0  
2          1478000.0  4735000.0  
3          1400000.0  5600000.0  
4          1298750.0  5500000.0  
5          1371250.0  8000000.0  
6          1325000.0  7650000.0  
7          1150000.0  9000000.0  
8          1310000.0  6500000.0  
9          1360000.0  6400000.0  
10         1416500.0  6250000.0  
11         1399125.0  5050000.0  
12         1412000.0  5700000.0

In [18]:
melb_df.groupby('Regionname')['SellerG'].agg(['nunique', set])

nunique  \
Regionname                            
Eastern Metropolitan             26   
Eastern Victoria                 11   
Northern Metropolitan            40   
Northern Victoria                11   
South-Eastern Metropolitan       25   
Southern Metropolitan            38   
Western Metropolitan             34   
Western Victoria                  6   

                                                                          set  
Regionname                                                                     
Eastern Metropolitan        {McGrath, Fletchers, Buckingham, HAR, Philip, ...  
Eastern Victoria            {McGrath, Fletchers, Harcourts, HAR, O'Brien, ...  
Northern Metropolitan       {McGrath, Fletchers, HAR, Buckingham, Philip, ...  
Northern Victoria           {McGrath, LITTLE, McDonald, Buckingham, HAR, Y...  
South-Eastern Metropolitan  {McGrath, Fletchers, HAR, hockingstuart, Ray, ...  
Southern Metropolitan       {McGrath, Fletchers, HAR, Buckingham, Philip, ...  
Western Metropolitan        {McGrath, HAR, hockingstuart, Jas, Ray, other,...  
Western Victoria                 {HAR, YPA, hockingstuart, Ray, Raine, other}

----

###  Задание 3.1 ###

Сгруппировать данные по признаку количества комнат и найти среднюю цену объектов недвижимости в каждой группе. В качестве ответа записать количество комнат, для которых средняя цена наибольшая.

***Ответ - 7***

In [19]:
display(melb_df.groupby('Rooms')['Price'].mean().max())
melb_df.groupby('Rooms')['Price'].mean()

1920700.0

Rooms
1     4.338245e+05
2     7.750812e+05
3     1.076081e+06
4     1.445282e+06
5     1.870260e+06
6     1.849366e+06
7     1.920700e+06
8     1.602750e+06
10    9.000000e+05
Name: Price, dtype: float64

----

###  Задание 3.2 ###

Какой регион имеет наименьшее стандартное отклонение по географической широте (Lattitude)?
В качестве ответа записать название этого региона.

***Ответ - Western Victoria***

In [20]:
display(melb_df.groupby('Regionname')['Lattitude'].std().min())
melb_df.groupby('Regionname')['Lattitude'].std()

0.011579033453868534

Regionname
Eastern Metropolitan          0.047890
Eastern Victoria              0.147067
Northern Metropolitan         0.049639
Northern Victoria             0.084455
South-Eastern Metropolitan    0.073411
Southern Metropolitan         0.043080
Western Metropolitan          0.051251
Western Victoria              0.011579
Name: Lattitude, dtype: float64

----

###  Задание 3.3 ###

Какая риелторская компания (`SellerG`) имеет наименьшую общую выручку за период с 1 мая по 1 сентября (включительно) 2017 года?

Для ответа на этот вопрос рассчитать сумму продаж (`Price`) каждой компании в заданный период.

Не забыть перевести даты в формат `datetime`.

***Ответ - LITTLE***

In [21]:
melb_df['Date'] = pd.to_datetime(melb_df['Date'], dayfirst=True)
begin_mask = melb_df['Date'] >= '2017-05-01'
end_mask = melb_df['Date'] <= '2017-09-01'
display(melb_df[begin_mask & end_mask].groupby('SellerG')['Price'].sum().sort_values())

SellerG
LITTLE             2742000.0
Cayzer             4439000.0
Burnham            4550500.0
Moonee             7328000.0
Thomson            8332000.0
Bells              8656000.0
Alexkarbon        10985000.0
McDonald          14637500.0
Rendina           15422276.0
Nick              16890000.0
Douglas           18341000.0
Buckingham        19033000.0
C21               19515000.0
Eview             19791500.0
Collins           20217000.0
Philip            22051800.0
Chisholm          23225000.0
Williams          23297000.0
Love              23365500.0
Purplebricks      23401000.0
O'Brien           23855508.0
HAR               25568000.0
Village           26473000.0
RW                29261000.0
Raine             30687700.0
Stockdale         35409800.0
Sweeney           36882750.0
Gary              39138400.0
Hodges            43231000.0
YPA               46354350.0
Miles             47582000.0
Kay               48569500.0
RT                50498000.0
Brad              55955000.0
Jas   

# Сводные таблицы #

## Метод groupby как способ построения сводных таблиц ##

In [22]:
melb_df.groupby('Rooms')[['Price', 'BuildingArea']].median()

Price  BuildingArea
Rooms                         
1       385000.0         107.0
2       690000.0         126.0
3       950000.0         126.0
4      1285000.0         142.0
5      1660000.0         176.0
6      1800000.0         126.0
7      1496000.0         216.5
8      1515000.0         126.0
10      900000.0         126.0

In [23]:
melb_df.groupby(['Rooms', 'Type'])['Price'].mean()

Rooms  Type     
1      house        8.668655e+05
       townhouse    5.927045e+05
       unit         3.899289e+05
2      house        1.017238e+06
       townhouse    7.101585e+05
       unit         6.104905e+05
3      house        1.109233e+06
       townhouse    9.847087e+05
       unit         8.505963e+05
4      house        1.462283e+06
       townhouse    1.217092e+06
       unit         1.037476e+06
5      house        1.877327e+06
       townhouse    1.035000e+06
       unit                  NaN
6      house        1.869508e+06
       townhouse             NaN
       unit         5.200000e+05
7      house        1.920700e+06
       townhouse             NaN
       unit                  NaN
8      house        1.510286e+06
       townhouse             NaN
       unit         2.250000e+06
10     house        9.000000e+05
       townhouse             NaN
       unit                  NaN
Name: Price, dtype: float64

In [24]:
melb_df.groupby(['Rooms', 'Type'])['Price'].mean().unstack()

Type          house     townhouse          unit
Rooms                                          
1      8.668655e+05  5.927045e+05  3.899289e+05
2      1.017238e+06  7.101585e+05  6.104905e+05
3      1.109233e+06  9.847087e+05  8.505963e+05
4      1.462283e+06  1.217092e+06  1.037476e+06
5      1.877327e+06  1.035000e+06           NaN
6      1.869508e+06           NaN  5.200000e+05
7      1.920700e+06           NaN           NaN
8      1.510286e+06           NaN  2.250000e+06
10     9.000000e+05           NaN           NaN

## Метод pivot_table для построения сводных таблиц ##

In [25]:
melb_df.pivot_table(
    values='Price',
    index='Rooms',
    columns='Type',
    fill_value=0
).round()

Type       house  townhouse       unit
Rooms                                 
1       866866.0   592705.0   389929.0
2      1017238.0   710158.0   610491.0
3      1109233.0   984709.0   850596.0
4      1462283.0  1217092.0  1037476.0
5      1877327.0  1035000.0        0.0
6      1869508.0        0.0   520000.0
7      1920700.0        0.0        0.0
8      1510286.0        0.0  2250000.0
10      900000.0        0.0        0.0

In [26]:
melb_df.pivot_table(
    values='Price',
    index='Regionname',
    columns='Weekend',
    aggfunc='count'
)

Weekend                        0     1
Regionname                            
Eastern Metropolitan         447  1024
Eastern Victoria              13    40
Northern Metropolitan       1258  2632
Northern Victoria             11    30
South-Eastern Metropolitan   123   327
Southern Metropolitan       1534  3161
Western Metropolitan         960  1988
Western Victoria               8    24

In [27]:
melb_df.pivot_table(
    values='Landsize',
    index='Regionname',
    columns='Type',
    aggfunc=['median', 'mean'],
    fill_value=0
)

median                        mean              \
Type                        house townhouse unit        house   townhouse   
Regionname                                                                  
Eastern Metropolitan        674.0     233.5  203   717.422847  269.440678   
Eastern Victoria            843.0       0.0  230  3108.960000    0.000000   
Northern Metropolitan       459.5     134.0    0   619.249092  317.325733   
Northern Victoria           724.0       0.0    0  3355.463415    0.000000   
South-Eastern Metropolitan  630.5     240.0  199   664.306701  212.160000   
Southern Metropolitan       586.0     246.0    0   569.643881  278.858824   
Western Metropolitan        531.0     198.0   62   507.883406  244.560669   
Western Victoria            599.5       0.0    0   655.500000    0.000000   

                                        
Type                              unit  
Regionname                              
Eastern Metropolitan        330.444444  
Eastern Victoria            295.333333  
Northern Metropolitan       495.026538  
Northern Victoria             0.000000  
South-Eastern Metropolitan  357.864865  
Southern Metropolitan       466.380245  
Western Metropolitan        557.637232  
Western Victoria              0.000000

## Многомерные сводные таблицы ##

In [28]:
melb_df.pivot_table(
    values='Price',
    index=['Method', 'Type'],
    columns='Regionname',
    aggfunc='median',
    fill_value=0
)

Regionname        Eastern Metropolitan  Eastern Victoria  \
Method Type                                                
PI     house                   1244000            780000   
       townhouse                760000                 0   
       unit                     650000                 0   
S      house                   1127000            675000   
       townhouse                828000                 0   
       unit                     645750            492000   
SA     house                    932500            950000   
       townhouse                807500                 0   
       unit                          0                 0   
SP     house                   1050000            672500   
       townhouse                910000                 0   
       unit                     515000            400000   
VB     house                   1100000            712500   
       townhouse                892500                 0   
       unit                     500000                 0   

Regionname        Northern Metropolitan  Northern Victoria  \
Method Type                                                  
PI     house                     900000             500000   
       townhouse                 632500                  0   
       unit                      410000                  0   
S      house                     920000             555000   
       townhouse                 750000                  0   
       unit                      525500                  0   
SA     house                     817500             540000   
       townhouse                 425000                  0   
       unit                      616000                  0   
SP     house                     900000             521000   
       townhouse                 690000                  0   
       unit                      470000                  0   
VB     house                    1050000             690000   
       townhouse                 640000                  0   
       unit                      450000                  0   

Regionname        South-Eastern Metropolitan  Southern Metropolitan  \
Method Type                                                           
PI     house                          865000                1725000   
       townhouse                     1190000                1055000   
       unit                           525000                 571250   
S      house                          883300                1611000   
       townhouse                      875000                1135000   
       unit                           606000                 655000   
SA     house                          880000                1390000   
       townhouse                           0                1141000   
       unit                                0                 580000   
SP     house                          770000                1521750   
       townhouse                      800000                1162500   
       unit                           601000                 550000   
VB     house                          850000                1800000   
       townhouse                           0                1250000   
       unit                           700000                 500000   

Regionname        Western Metropolitan  Western Victoria  
Method Type                                               
PI     house                    870000            630000  
       townhouse                670000                 0  
       unit                     360000                 0  
S      house                    870000            397500  
       townhouse                729000                 0  
       unit                     489000                 0  
SA     house                    772500                 0  
       townhouse                467500                 0  
       unit                     571000                 0  
SP     house                    865000            360000  
       t

## Доступ к данным в сводной таблице ##

In [29]:
pivot = melb_df.pivot_table(
    values='Landsize',
    index='Regionname',
    columns='Type',
    aggfunc=['median', 'mean'],
    fill_value=0
)
pivot.columns

MultiIndex([('median',     'house'),
            ('median', 'townhouse'),
            ('median',      'unit'),
            (  'mean',     'house'),
            (  'mean', 'townhouse'),
            (  'mean',      'unit')],
           names=[None, 'Type'])

In [30]:
display(pivot)
display(pivot['mean'])
display(pivot['mean']['unit'])

median                        mean              \
Type                        house townhouse unit        house   townhouse   
Regionname                                                                  
Eastern Metropolitan        674.0     233.5  203   717.422847  269.440678   
Eastern Victoria            843.0       0.0  230  3108.960000    0.000000   
Northern Metropolitan       459.5     134.0    0   619.249092  317.325733   
Northern Victoria           724.0       0.0    0  3355.463415    0.000000   
South-Eastern Metropolitan  630.5     240.0  199   664.306701  212.160000   
Southern Metropolitan       586.0     246.0    0   569.643881  278.858824   
Western Metropolitan        531.0     198.0   62   507.883406  244.560669   
Western Victoria            599.5       0.0    0   655.500000    0.000000   

                                        
Type                              unit  
Regionname                              
Eastern Metropolitan        330.444444  
Eastern Victoria            295.333333  
Northern Metropolitan       495.026538  
Northern Victoria             0.000000  
South-Eastern Metropolitan  357.864865  
Southern Metropolitan       466.380245  
Western Metropolitan        557.637232  
Western Victoria              0.000000

Type                              house   townhouse        unit
Regionname                                                     
Eastern Metropolitan         717.422847  269.440678  330.444444
Eastern Victoria            3108.960000    0.000000  295.333333
Northern Metropolitan        619.249092  317.325733  495.026538
Northern Victoria           3355.463415    0.000000    0.000000
South-Eastern Metropolitan   664.306701  212.160000  357.864865
Southern Metropolitan        569.643881  278.858824  466.380245
Western Metropolitan         507.883406  244.560669  557.637232
Western Victoria             655.500000    0.000000    0.000000

Regionname
Eastern Metropolitan          330.444444
Eastern Victoria              295.333333
Northern Metropolitan         495.026538
Northern Victoria               0.000000
South-Eastern Metropolitan    357.864865
Southern Metropolitan         466.380245
Western Metropolitan          557.637232
Western Victoria                0.000000
Name: unit, dtype: float64

In [31]:
mask = pivot['mean']['house'] < pivot['median']['house']
filtered_pivot = pivot[mask]
display(filtered_pivot)

print(list(filtered_pivot.index))

median                       mean              \
Type                   house townhouse unit       house   townhouse   
Regionname                                                            
Southern Metropolitan  586.0     246.0    0  569.643881  278.858824   
Western Metropolitan   531.0     198.0   62  507.883406  244.560669   

                                   
Type                         unit  
Regionname                         
Southern Metropolitan  466.380245  
Western Metropolitan   557.637232

['Southern Metropolitan', 'Western Metropolitan']


In [32]:
import numpy as np

mser = pd.Series(
    np.random.rand(8),
    index=[['white', 'white', 'white', 'blue', 'blue', 'red', 'red', 'red'],
           ['up', 'down', 'right', 'up', 'down', 'up', 'down', 'left']
          ]
)
display(mser)

white  up       0.975780
       down     0.139964
       right    0.511932
blue   up       0.174798
       down     0.705015
red    up       0.197045
       down     0.204700
       left     0.901609
dtype: float64

In [33]:
print(mser.index)

MultiIndex([('white',    'up'),
            ('white',  'down'),
            ('white', 'right'),
            ( 'blue',    'up'),
            ( 'blue',  'down'),
            (  'red',    'up'),
            (  'red',  'down'),
            (  'red',  'left')],
           )


In [34]:
mframe = pd.DataFrame(
    np.random.randn(16).reshape(4, 4),
    index=[['white', 'white', 'red', 'red'],
           ['up', 'down', 'up', 'down']],
    columns=[['pen', 'pen', 'paper', 'paper'],
             [1, 2, 1, 2]]
)
display(mframe)

pen               paper          
                   1         2         1         2
white up   -1.371463  0.974635 -0.375429  1.956725
      down  1.426371  1.208112 -0.263676  0.980558
red   up    0.370294 -0.177660  0.835687 -0.969236
      down -0.172826 -0.575394  2.047861 -0.051283

----

###  Задание 4.1 ###

Какой параметр метода `pivot_table()` отвечает за признак, по которому будут рассчитаны агрегирующие функции?

***Ответ - values***

----

###  Задание 4.2 ###

Составить сводную таблицу, которая показывает зависимость медианной площади (`BuildingArea`) здания от типа объекта недвижимости (`Type`) и количества жилых комнат в доме (`Rooms`). Для какой комбинации признаков площадь здания наибольшая?

В качестве ответа записать эту комбинацию (тип здания, число комнат) через запятую, без пробелов.

***Ответ - house,7***

In [35]:
melb_df.groupby(['Type', 'Rooms'])['BuildingArea'].median().sort_values(ascending=False)

Type       Rooms
house      7        216.5
           5        177.0
unit       6        171.0
townhouse  4        159.5
           5        152.0
house      4        141.0
           1        126.0
townhouse  3        126.0
unit       4        126.0
           3        126.0
           8        126.0
house      2        126.0
           10       126.0
           8        126.0
           6        126.0
           3        126.0
townhouse  2        114.0
unit       2        110.0
townhouse  1         88.0
unit       1         69.5
townhouse  6          NaN
           7          NaN
           8          NaN
           10         NaN
unit       5          NaN
           7          NaN
           10         NaN
Name: BuildingArea, dtype: float64

In [36]:
melb_df.pivot_table(
    index=['Type', 'Rooms'],
    values='BuildingArea',
    aggfunc='median',
    fill_value=0
).sort_values(by='BuildingArea', ascending=False)

BuildingArea
Type      Rooms              
house     7             216.5
          5             177.0
unit      6             171.0
townhouse 4             159.5
          5             152.0
house     4             141.0
          1             126.0
townhouse 3             126.0
unit      4             126.0
          3             126.0
          8             126.0
house     2             126.0
          10            126.0
          8             126.0
          6             126.0
          3             126.0
townhouse 2             114.0
unit      2             110.0
townhouse 1              88.0
unit      1              69.5

----

###  Задание 4.3 ###

Составить сводную таблицу, которая показывает зависимость медианной цены объекта недвижимости (`Price`) от риелторского агентства (`SellerG`) и типа здания (`Type`).

Во вновь созданной таблице найти агентство, у которого медианная цена для зданий типа `unit` максимальна. В качестве ответа записать название этого агентства.

***Ответ - Nick***

In [37]:
melb_df.pivot_table(
    values='Price',
    columns='Type',
    index='SellerG',
    aggfunc='median'
)['unit'].sort_values(ascending=False)

SellerG
Nick             900000.0
Marshall         715000.0
Cayzer           707500.0
Kay              695000.0
Noel             693250.0
Buxton           670000.0
Fletchers        653000.0
Chisholm         640000.0
Philip           636000.0
RT               630000.0
C21              625500.0
RW               625500.0
Moonee           622000.0
O'Brien          618500.0
Hodges           605500.0
Eview            602000.0
Woodards         600000.0
Collins          592000.0
Jellis           591750.0
Purplebricks     582500.0
Williams         577000.0
hockingstuart    566000.0
Alexkarbon       560000.0
Miles            555000.0
Buckingham       553000.0
Gary             550000.0
Biggin           548000.0
Greg             547500.0
McGrath          542500.0
other            535000.0
Nelson           530000.0
HAR              518000.0
McDonald         502500.0
Barry            502000.0
Harcourts        500000.0
Ray              497500.0
Thomson          495000.0
Douglas          494500.0
Brad

# Объединение DataFrame: знакомимся с новыми данными #

----

###  Задание 5.1 ###

Значения из какого столбца таблиц `ratings1` и `ratings2` можно расшифровать с помощью таблицы `movies`?

***Ответ - movieId***

In [52]:
ratings1_sample = pd.read_csv('data/ratings1.csv')
ratings2_sample = pd.read_csv('data/ratings2.csv')
movies_sample = pd.read_csv('data/movies.csv')
display(ratings1_sample)
display(ratings1_sample)
display(movies_sample)

userId  movieId  rating
0           1        1     4.0
1           1        3     4.0
2           1        6     4.0
3           1       47     5.0
4           1       50     5.0
...       ...      ...     ...
39996     274     5582     2.5
39997     274     5585     3.5
39998     274     5588     3.5
39999     274     5618     4.0
40000     274     5621     2.0

[40001 rows x 3 columns]

userId  movieId  rating
0           1        1     4.0
1           1        3     4.0
2           1        6     4.0
3           1       47     5.0
4           1       50     5.0
...       ...      ...     ...
39996     274     5582     2.5
39997     274     5585     3.5
39998     274     5588     3.5
39999     274     5618     4.0
40000     274     5621     2.0

[40001 rows x 3 columns]

movieId                                      title  \
0           1                           Toy Story (1995)   
1           2                             Jumanji (1995)   
2           3                    Grumpier Old Men (1995)   
3           4                   Waiting to Exhale (1995)   
4           5         Father of the Bride Part II (1995)   
...       ...                                        ...   
9737   193581  Black Butler: Book of the Atlantic (2017)   
9738   193583               No Game No Life: Zero (2017)   
9739   193585                               Flint (2017)   
9740   193587        Bungo Stray Dogs: Dead Apple (2018)   
9741   193609        Andrew Dice Clay: Dice Rules (1991)   

                                           genres  
0     Adventure|Animation|Children|Comedy|Fantasy  
1                      Adventure|Children|Fantasy  
2                                  Comedy|Romance  
3                            Comedy|Drama|Romance  
4                                          Comedy  
...                                           ...  
9737              Action|Animation|Comedy|Fantasy  
9738                     Animation|Comedy|Fantasy  
9739                                        Drama  
9740                             Action|Animation  
9741                                       Comedy  

[9742 rows x 3 columns]

----

###  Задание 5.2 ###

Сколько уникальных фильмов представлено в таблице `movies`?

***Ответ - 9742***

In [41]:
movies_sample.nunique()

movieId    9742
title      9737
genres      951
dtype: int64

----

###  Задание 5.3 ###

Сколько уникальных пользователей в таблице `ratings1`?

***Ответ - 274***

In [53]:
ratings1_sample.nunique()

userId      274
movieId    6219
rating       10
dtype: int64

----

###  Задание 5.4 ###

В каком году было выставлено больше всего оценок?

Для ответа на этот вопрос использовать таблицу `dates`.

***Ответ - 2000***

In [54]:
dates_sample = pd.read_csv('data/dates.csv')
dates_sample['date'] = pd.to_datetime(dates_sample['date'])
display(dates_sample.info())
dates_sample['year'] = dates_sample['date'].dt.year
display(dates_sample['year'].value_counts())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100836 entries, 0 to 100835
Data columns (total 1 columns):
 #   Column  Non-Null Count   Dtype         
---  ------  --------------   -----         
 0   date    100836 non-null  datetime64[ns]
dtypes: datetime64[ns](1)
memory usage: 787.9 KB


None

2000    10061
2017     8198
2007     7114
2016     6703
2015     6616
2018     6418
1996     6040
2005     5813
2012     4656
2008     4351
2009     4158
2006     4059
2003     4014
2001     3922
2002     3478
2004     3279
1999     2439
2010     2301
1997     1916
2011     1690
2013     1664
2014     1439
1998      507
Name: year, dtype: int64

# Объединение DataFrame: concat #